pkg/flow/internal/controller/node_config_import.go

```diff
@@ -0,0 +1,425 @@
+
+	runner := runner.New(func(node *ImportConfigNode) runner.Worker {
+		return &childRunner{
+			node:    node,
+			errChan: errChan,
+		}
+	})
+	defer runner.Stop()
+
+	updateTasks := func() error {
+		cn.mut.Lock()
+		defer cn.mut.Unlock()
+		cn.importChildrenRunning = true
+		var tasks []*ImportConfigNode
+		for _, value := range cn.importConfigNodesChildren {
+			tasks = append(tasks, value)
+		}
+
+		return runner.ApplyTasks(newCtx, tasks)
+	}
+
+	err := updateTasks()
+	if err != nil {
+		level.Error(cn.logger).Log("msg", "import failed to run nested imports", "err", err)
+		cn.setRunHealth(component.HealthTypeExited, fmt.Sprintf("import shut down while trying to run nested imports with error: %s", err))
+		return err
+	}
+
+	go func() {
+		errChan <- cn.source.Run(newCtx)
+	}()
+
+	cn.setRunHealth(component.HealthTypeHealthy, "started import")
+	err = cn.run(errChan, updateTasks)
+
+	var exitMsg string
+	if err != nil {
+		level.Error(cn.logger).Log("msg", "import exited with error", "err", err)
+		exitMsg = fmt.Sprintf("import shut down with error: %s", err)
+	} else {
+		level.Info(cn.logger).Log("msg", "import exited")
+		exitMsg = "import shut down normally"
+	}
+	cn.setRunHealth(component.HealthTypeExited, exitMsg)
+	return err
+}
+
+func (cn *ImportConfigNode) run(errChan chan error, updateTasks func() error) error {
+	for {
+		select {
+		case <-cn.importChildrenUpdateChan:
+			err := updateTasks()
+			if err != nil {
+				return err
+			}
+		case err := <-errChan:
+			return err
+		}
+	}
+}
+
+func (cn *ImportConfigNode) Label() string { return cn.label }
+
+// Block implements BlockNode and returns the current block of the managed config node.
+func (cn *ImportConfigNode) Block() *ast.BlockStmt { return cn.block }
+
+// NodeID implements dag.Node and returns the unique ID for the config node.
+func (cn *ImportConfigNode) NodeID() string { return cn.nodeID }
+
+// ImportedDeclares returns all declare blocks that it imported.
+func (cn *ImportConfigNode) ImportedDeclares() map[string]ast.Body {
+	cn.mut.RLock()
+	defer cn.mut.RUnlock()
+	return cn.importedDeclares
+}
+
+// ImportConfigNodesChildren returns the ImportConfigNodesChildren of this ImportConfigNode.
+func (cn *ImportConfigNode) ImportConfigNodesChildren() map[string]*ImportConfigNode {
+	cn.mut.Lock()
+	defer cn.mut.Unlock()
+	return cn.importConfigNodesChildren
+}
+
+type childRunner struct {
+	node    *ImportConfigNode
+	errChan chan error
+}
+
+func (cr *childRunner) Run(ctx context.Context) {
+	err := cr.node.Run(ctx)
+	if err != nil {
+		cr.errChan <- err
+	}
+}
+
+// TODO: is nodeID enough?
+func (cn *ImportConfigNode) Hash() uint64 {
+	fnvHash := fnv.New64a()
+	fnvHash.Write([]byte(cn.NodeID()))
+	return fnvHash.Sum64()
+}
+
+// TODO: is nodeID enough?
+func (cn *ImportConfigNode) Equals(other runner.Task) bool {
+	otherTask := other.(*ImportConfigNode)
+
+	if cn == otherTask {
+		return true
+	}
+
+	return cn.NodeID() == otherTask.NodeID()
+}
```

Comment:
Yes; nodeID should be enough in both cases here. `Hash` and `Equals` are used to find identical runner tasks; since node IDs are unique per graph, it is safe to use it for hashing and equality checking here. 

(Also, the runner package handles hash collision gracefully, so it's OK if the hash is imperfect)

Reply:
There is actually a problem with this:
- root imports "module" that imports "nested_module_1"
- "module" is updated to import "nested_module_2" instead, but the nodeID of the import config block is unchanged
- the runner pkg will think that the exact same task is already running (because the nodeID is the same) and will ignore the task

=> the runner will keep running the old child and the new child won't run

In our case I believe that we don't want this optimization, we always want to stop the previous workers and start new ones because we are creating new nodes.

the easy way to do this would be to have Equals return false (hash doesn't matter anyway because if there is a collision it will check with Equals)

I'm not worried about performance because module updates are very rare (there are checks to make sure that if the imported content is the same we don't do anything, that means that we would need a real change to update the children)

wdyt?

Reply:
(we also don't have the risk of running twice the same task because we check that the imported blocks are not duplicated)

Reply:
@wildum Normally we'd expect nodes to be dynamically updated instead of recycled/replaced, but let's change `Equals` to be a more basic pointer check for now and we can revisit this later once the refactorings for the controller start coming in.